File: backend/app/api/schemas/financas.py
"""Schemas do Organizador Financeiro pessoal (domínio `financas`).

Isolado dos schemas da Reative (prospector/copywriter/outreach) e dos
pessoais (perfil/vagas).
"""
from __future__ import annotations

from datetime import date
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel, Field


# ══════════════════════════════════════════════════════════════════
# Contas (onde o dinheiro mora)
# ══════════════════════════════════════════════════════════════════

class ContaCreate(BaseModel):
    usuario_id: str = Field(..., description="Perfil dono da conta (você/Sandra)")
    nome: str = Field(..., description='ex: "Nubank", "Carteira", "VR Caju"')
    tipo: str = Field(..., description="corrente/dinheiro/vr/va/reserva/cartao_credito")
    saldo_atual: Decimal = Field(
        Decimal("0"), description="Saldo inicial (de abertura). Default 0."
    )


class ContaUpdate(BaseModel):
    nome: Optional[str] = None
    tipo: Optional[str] = None
    ativa: Optional[bool] = None


class ContaResponse(BaseModel):
    id: str
    usuario_id: str
    nome: str
    tipo: str
    saldo_atual: Decimal
    ativa: bool
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class ContaListResponse(BaseModel):
    items: List[ContaResponse]
    total: int


# ══════════════════════════════════════════════════════════════════
# Categorias (hierárquicas, compartilhadas)
# ══════════════════════════════════════════════════════════════════

class CategoriaCreate(BaseModel):
    nome: str
    categoria_pai_id: Optional[str] = Field(
        None, description="Pai na hierarquia. Null = categoria raiz."
    )


class CategoriaUpdate(BaseModel):
    nome: Optional[str] = None
    categoria_pai_id: Optional[str] = None
    ativa: Optional[bool] = None


class CategoriaResponse(BaseModel):
    id: str
    nome: str
    categoria_pai_id: Optional[str] = None
    ativa: bool
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class CategoriaTreeItem(BaseModel):
    """Nó da árvore de categorias (pai com seus filhos aninhados)."""
    id: str
    nome: str
    ativa: bool
    filhos: List["CategoriaTreeItem"] = Field(default_factory=list)


class CategoriaTreeResponse(BaseModel):
    items: List[CategoriaTreeItem]   # raízes
    total: int                       # total de categorias (todos os níveis)


CategoriaTreeItem.model_rebuild()


# ══════════════════════════════════════════════════════════════════
# Transações
# ══════════════════════════════════════════════════════════════════

class DespesaCreate(BaseModel):
    """Lançamento de despesa simples (uma conta, uma categoria opcional)."""
    usuario_id: str
    descricao: str
    valor_total: Decimal = Field(..., gt=0)
    conta_id: str = Field(..., description="Conta de onde o dinheiro saiu")
    categoria_id: Optional[str] = None
    data_competencia: Optional[date] = Field(
        None, description="Mês de competência. Default: hoje."
    )
    data_pagamento: Optional[date] = Field(
        None, description="Quando saiu de fato. Default: hoje se status=paga."
    )
    status: str = Field("paga", description="prevista/paga/atrasada")
    notas: Optional[str] = None


class PagamentoIn(BaseModel):
    conta_id: str
    valor: Decimal = Field(..., gt=0)


class DespesaDivididaCreate(BaseModel):
    """Despesa paga por N contas (split explícito). A soma dos pagamentos
    precisa bater com valor_total."""
    usuario_id: str
    descricao: str
    valor_total: Decimal = Field(..., gt=0)
    pagamentos: List[PagamentoIn] = Field(..., min_length=1)
    categoria_id: Optional[str] = None
    data_competencia: Optional[date] = None
    data_pagamento: Optional[date] = None
    status: str = "paga"
    notas: Optional[str] = None


class DespesaAutoSplitCreate(BaseModel):
    """Despesa que esgota o VR/VA e joga o resto no dinheiro automaticamente.
    Resolve o 'às vezes acaba o VR'. Sempre lançada como paga."""
    usuario_id: str
    descricao: str
    valor_total: Decimal = Field(..., gt=0)
    conta_vr_id: str = Field(..., description="Conta que esgota primeiro (VR/VA)")
    conta_fallback_id: str = Field(..., description="Conta que cobre o resto (dinheiro)")
    categoria_id: Optional[str] = None
    data_competencia: Optional[date] = None
    notas: Optional[str] = None


class ReceitaCreate(BaseModel):
    """Lançamento de receita simples (entra dinheiro numa conta)."""
    usuario_id: str
    descricao: str
    valor_total: Decimal = Field(..., gt=0)
    conta_id: str = Field(..., description="Conta que recebeu o dinheiro")
    categoria_id: Optional[str] = None
    data_competencia: Optional[date] = None
    data_pagamento: Optional[date] = None
    status: str = Field("paga", description="prevista/paga/atrasada")
    notas: Optional[str] = None


class TransacaoUpdate(BaseModel):
    """Edição de uma transação simples (uma conta). Espelha o formulário do
    dashboard: tipo, descrição, valor, conta, categoria, data e status. O saldo
    da conta é reajustado (reverte o efeito antigo e aplica o novo)."""
    tipo: str = Field(..., description="despesa | receita")
    descricao: str
    valor_total: Decimal = Field(..., gt=0)
    conta_id: str = Field(..., description="Conta da transação")
    categoria_id: Optional[str] = None
    data_competencia: Optional[date] = None
    status: str = Field("paga", description="prevista/paga/atrasada")


class PagarTransacaoRequest(BaseModel):
    """Quita uma transação prevista/atrasada (move o saldo). ``conta_id`` só é
    exigido quando a transação ainda não tem conta (boleto importado /
    recorrência); se já tem pagamento(s), efetiva nas contas existentes."""
    conta_id: Optional[str] = None
    data_pagamento: Optional[date] = None


class TransacaoItemResponse(BaseModel):
    id: str
    categoria_id: Optional[str] = None
    descricao: str
    valor: Decimal


class TransacaoPagamentoResponse(BaseModel):
    id: str
    conta_id: str
    valor: Decimal


class TransacaoResponse(BaseModel):
    id: str
    usuario_id: str
    tipo: str
    descricao: str
    valor_total: Decimal
    data_competencia: date
    data_pagamento: Optional[date] = None
    data_vencimento: Optional[date] = None
    status: str
    origem: str
    categoria_id: Optional[str] = None
    notas: Optional[str] = None
    itens: List[TransacaoItemResponse] = Field(default_factory=list)
    pagamentos: List[TransacaoPagamentoResponse] = Field(default_factory=list)
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class TransacaoListItem(BaseModel):
    """Linha enxuta para a lista filtrável do dashboard (sem itens/pagamentos
    completos — só os nomes das contas pagadoras e da categoria)."""
    id: str
    tipo: str
    descricao: str
    valor_total: Decimal
    data_competencia: date
    data_pagamento: Optional[date] = None
    data_vencimento: Optional[date] = None
    status: str
    categoria_id: Optional[str] = None
    categoria_nome: Optional[str] = None
    contas: List[str] = Field(default_factory=list)


class TransacaoListResponse(BaseModel):
    items: List[TransacaoListItem]
    total: int
    limit: int
    offset: int


# ══════════════════════════════════════════════════════════════════
# Cartões, faturas, compras parceladas e parcelas
# ══════════════════════════════════════════════════════════════════

class CartaoCreate(BaseModel):
    usuario_id: str
    nome: str
    bandeira: Optional[str] = None
    dia_fechamento: int = Field(..., ge=1, le=31)
    dia_vencimento: int = Field(..., ge=1, le=31)
    limite: Optional[Decimal] = Field(None, ge=0)


class CartaoUpdate(BaseModel):
    nome: Optional[str] = None
    bandeira: Optional[str] = None
    dia_fechamento: Optional[int] = Field(None, ge=1, le=31)
    dia_vencimento: Optional[int] = Field(None, ge=1, le=31)
    limite: Optional[Decimal] = Field(None, ge=0)
    ativo: Optional[bool] = None


class CartaoResponse(BaseModel):
    id: str
    usuario_id: str
    nome: str
    bandeira: Optional[str] = None
    dia_fechamento: int
    dia_vencimento: int
    limite: Optional[Decimal] = None
    ativo: bool
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class CartaoListResponse(BaseModel):
    items: List["CartaoResponse"]
    total: int


class FaturaResponse(BaseModel):
    id: str
    cartao_id: str
    mes_referencia: date
    valor_total: Decimal
    vencimento: date
    status: str


class FaturasCartaoResponse(BaseModel):
    cartao_id: str
    faturas: List[FaturaResponse]
    total_em_aberto: Decimal       # soma das faturas não pagas
    total_juros: Decimal           # soma de valor_juros das parcelas do cartão


class CompraParceladaCreate(BaseModel):
    """Compra no cartão parcelada em N vezes."""
    usuario_id: str
    cartao_id: str
    descricao: str
    valor_total: Decimal = Field(..., gt=0, description="Total a pagar (já com juros)")
    total_parcelas: int = Field(..., ge=1, le=120)
    data_compra: Optional[date] = None
    categoria_id: Optional[str] = None
    valor_juros_total: Decimal = Field(
        Decimal("0"), ge=0, description="Quanto do total é juro (distribuído nas parcelas)"
    )


class BoletoParceladoCreate(BaseModel):
    """Despesa parcelada em boleto (sem cartão/fatura), ex.: reforma do
    condomínio em 6x. As parcelas vencem mês a mês a partir de primeiro_vencimento."""
    usuario_id: str
    descricao: str
    valor_total: Decimal = Field(..., gt=0)
    total_parcelas: int = Field(..., ge=1, le=120)
    primeiro_vencimento: date
    categoria_id: Optional[str] = None
    valor_juros_total: Decimal = Field(Decimal("0"), ge=0)


class ParcelaResponse(BaseModel):
    id: str
    numero: int
    total_parcelas: int
    valor: Decimal
    tem_juros: bool
    valor_juros: Decimal
    vencimento: date
    fatura_id: Optional[str] = None


class CompraResponse(BaseModel):
    id: str
    usuario_id: str
    cartao_id: Optional[str] = None
    descricao: str
    valor_total: Decimal
    total_parcelas: int
    data_compra: date
    origem: str
    categoria_id: Optional[str] = None
    parcelas: List[ParcelaResponse] = Field(default_factory=list)
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


# ══════════════════════════════════════════════════════════════════
# Recorrências (despesas/receitas fixas)
# ══════════════════════════════════════════════════════════════════

class RecorrenciaCreate(BaseModel):
    usuario_id: str
    descricao: str
    tipo: str = Field("despesa", description="despesa/receita")
    valor_estimado: Decimal = Field(..., gt=0)
    dia_vencimento: int = Field(..., ge=1, le=31)
    categoria_id: Optional[str] = None
    conta_id: Optional[str] = None
    frequencia: str = "mensal"


class RecorrenciaUpdate(BaseModel):
    descricao: Optional[str] = None
    tipo: Optional[str] = None
    valor_estimado: Optional[Decimal] = Field(None, gt=0)
    dia_vencimento: Optional[int] = Field(None, ge=1, le=31)
    categoria_id: Optional[str] = None
    conta_id: Optional[str] = None
    ativa: Optional[bool] = None


class RecorrenciaResponse(BaseModel):
    id: str
    usuario_id: str
    descricao: str
    tipo: str
    valor_estimado: Decimal
    dia_vencimento: int
    frequencia: str
    categoria_id: Optional[str] = None
    conta_id: Optional[str] = None
    ativa: bool
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class RecorrenciaListResponse(BaseModel):
    items: List[RecorrenciaResponse]
    total: int


class ProcessarRecorrenciasResponse(BaseModel):
    previstas_criadas: int
    marcadas_atrasadas: int


# ══════════════════════════════════════════════════════════════════
# NLU do Telegram (texto livre → estrutura)
# ══════════════════════════════════════════════════════════════════

class NLUResult(BaseModel):
    """Saída crua do LLM ao interpretar o texto livre."""
    tipo: str                              # despesa/receita
    valor: Decimal
    descricao: str
    categoria: Optional[str] = None        # nome (o service resolve pro id)
    conta: Optional[str] = None            # nome (o service resolve pro id)
    data: Optional[date] = None


class InterpretarTextoRequest(BaseModel):
    usuario_id: str
    texto: str


class InterpretacaoResponse(BaseModel):
    """Rascunho interpretado — o usuário confirma antes de virar transação."""
    tipo: str
    valor: Decimal
    descricao: str
    data: date
    conta_id: Optional[str] = None
    conta_nome: Optional[str] = None
    categoria_id: Optional[str] = None
    categoria_nome: Optional[str] = None
    texto_original: str


# ══════════════════════════════════════════════════════════════════
# Importador de boleto (LLM multimodal)
# ══════════════════════════════════════════════════════════════════

class VerbaBoleto(BaseModel):
    descricao: str
    valor: Decimal


class LeituraBoleto(BaseModel):
    tipo: str                              # agua/gas/luz
    leitura_atual: Optional[Decimal] = None
    leitura_anterior: Optional[Decimal] = None
    consumo: Optional[Decimal] = None
    valor: Optional[Decimal] = None


class BoletoExtraido(BaseModel):
    beneficiario: Optional[str] = None
    vencimento: Optional[date] = None
    valor_total: Decimal
    verbas: List[VerbaBoleto] = Field(default_factory=list)
    leituras: List[LeituraBoleto] = Field(default_factory=list)


class ImportarBoletoResponse(BaseModel):
    success: bool                          # conseguiu ler o arquivo?
    conferido: bool                        # soma das verbas == total?
    mensagem: str
    comprovante_id: Optional[str] = None
    transacao_id: Optional[str] = None     # criada só quando conferido
    extraido: Optional[BoletoExtraido] = None


# ══════════════════════════════════════════════════════════════════
# Comprovantes (arquivos no MinIO)
# ══════════════════════════════════════════════════════════════════

class ComprovanteResponse(BaseModel):
    id: str
    usuario_id: str
    transacao_id: Optional[str] = None
    tipo: str
    bucket: str
    arquivo_path: str
    nome_original: Optional[str] = None
    content_type: Optional[str] = None
    tamanho: Optional[int] = None
    hash: str
    url: Optional[str] = None        # presigned, preenchida sob demanda
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class ComprovanteListResponse(BaseModel):
    items: List[ComprovanteResponse]
    total: int


# ══════════════════════════════════════════════════════════════════
# Leituras de consumo (água/gás/luz)
# ══════════════════════════════════════════════════════════════════

class LeituraConsumoCreate(BaseModel):
    usuario_id: str
    tipo: str = Field(..., description="agua/gas/luz")
    mes_referencia: date
    leitura_atual: Decimal
    leitura_anterior: Optional[Decimal] = None
    consumo: Optional[Decimal] = Field(
        None, description="Se omitido, calcula atual − anterior"
    )
    valor: Optional[Decimal] = None
    transacao_id: Optional[str] = None


class LeituraConsumoResponse(BaseModel):
    id: str
    usuario_id: str
    tipo: str
    mes_referencia: date
    leitura_atual: Decimal
    leitura_anterior: Optional[Decimal] = None
    consumo: Optional[Decimal] = None
    valor: Optional[Decimal] = None
    transacao_id: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class LeituraConsumoListResponse(BaseModel):
    items: List[LeituraConsumoResponse]
    total: int


# ══════════════════════════════════════════════════════════════════
# Resumo do mês
# ══════════════════════════════════════════════════════════════════

class CategoriaResumoItem(BaseModel):
    categoria_id: Optional[str] = None
    categoria_nome: str            # "Sem categoria" quando null
    total: Decimal


class ResumoMesResponse(BaseModel):
    ano: int
    mes: int
    total_receitas: Decimal
    total_despesas: Decimal
    saldo: Decimal                 # receitas − despesas (sobra/déficit)
    por_categoria: List[CategoriaResumoItem]   # despesas, maior → menor


class RelatorioMesItem(BaseModel):
    """Um mês na série do relatório (cronológico, mais antigo → mais novo)."""
    ano: int
    mes: int
    total_receitas: Decimal
    total_despesas: Decimal
    saldo: Decimal                 # resultado do mês (receitas − despesas)


class RelatorioResponse(BaseModel):
    """Relatório do período (N meses até o mês âncora): série mês a mês,
    top categorias e totais consolidados."""
    meses: List[RelatorioMesItem]
    por_categoria: List[CategoriaResumoItem]   # despesas do período, maior → menor
    total_receitas: Decimal
    total_despesas: Decimal
    saldo: Decimal
    media_despesas: Decimal        # média mensal de despesa no período

File: frontend/lib/types/financas.ts
// Finanças (Organizador Financeiro pessoal).
//
// Domínio grande — candidato a sub-divisão futura (conta / transacao / cartao /
// recorrencia / orcamento / resumo / boleto). Ver docs/ORGANIZACAO_REFATORACAO.md.

// ── Contas ──────────────────────────────────────────────────────────
export interface Conta {
  id: string;
  usuario_id: string;
  nome: string;
  tipo: string;
  saldo_atual: string;
  meta?: string | null;
  ativa: boolean;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface ContaListResponse {
  items: Conta[];
  total: number;
}

/** Tipos de conta aceitos pelo backend (TIPOS_CONTA). */
export type TipoConta =
  | 'corrente'
  | 'dinheiro'
  | 'vr'
  | 'va'
  | 'reserva'
  | 'cartao_credito';

export interface ContaCreateInput {
  nome: string;
  tipo: TipoConta;
  saldo_atual?: string;
  meta?: string | null;
}

export interface ContaUpdateInput {
  nome?: string;
  tipo?: TipoConta;
  ativa?: boolean;
  meta?: string | null;
}

export interface CategoriaResumoItem {
  categoria_id?: string | null;
  categoria_nome: string;
  total: string;
}

// ── Transações (lista filtrável + lançamento pela web) ──────────────
export interface TransacaoListItem {
  id: string;
  tipo: string;
  descricao: string;
  valor_total: string;
  data_competencia: string;
  data_pagamento?: string | null;
  data_vencimento?: string | null;
  multa_percentual?: string | null;
  juros_mensal_percentual?: string | null;
  encargos_pagos?: string | null;
  linha_digitavel?: string | null;
  desconto_valor?: string | null;
  desconto_ate?: string | null;
  status: string;
  categoria_id?: string | null;
  categoria_nome?: string | null;
  contas: string[];
}

export interface TransacaoListResponse {
  items: TransacaoListItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface TransacaoPagamento {
  id: string;
  conta_id: string;
  valor: string;
}

/** Resposta de detalhe/criação de transação. O detalhe (GET /{id}) traz os
 *  `pagamentos`, usados pra pré-preencher a conta na edição. */
export interface TransacaoResponse {
  id: string;
  usuario_id: string;
  tipo: string;
  descricao: string;
  valor_total: string;
  data_competencia: string;
  data_pagamento?: string | null;
  data_vencimento?: string | null;
  multa_percentual?: string | null;
  juros_mensal_percentual?: string | null;
  encargos_pagos?: string | null;
  linha_digitavel?: string | null;
  desconto_valor?: string | null;
  desconto_ate?: string | null;
  status: string;
  categoria_id?: string | null;
  recorrencia_id?: string | null;
  pagamentos?: TransacaoPagamento[];
  itens?: { id: string; descricao: string; valor: string; categoria_id?: string | null }[];
}

export interface VerbaInput {
  descricao: string;
  valor: string;
}

/** Edição de uma conta a pagar (prevista) — não mexe no saldo. */
export interface PrevistaUpdateInput {
  descricao: string;
  valor_total: string;
  categoria_id?: string | null;
  data_vencimento?: string | null;
  multa_percentual?: string | null;
  juros_mensal_percentual?: string | null;
  itens?: VerbaInput[] | null;
  recorrencia_id?: string | null;
}

export interface TransacaoFiltro {
  ano?: number;
  mes?: number;
  conta_id?: string;
  categoria_id?: string;
  tipo?: 'despesa' | 'receita';
  /** prevista/paga/atrasada — pode passar várias (ex.: a pagar = previstas+atrasadas). */
  status?: string[];
  busca?: string;
  /** Ordena por vencimento (vencidas primeiro) — usado no painel "A pagar". */
  por_vencimento?: boolean;
  limit?: number;
  offset?: number;
}

/** Payload pra lançar despesa ou receita pela web. O `usuario_id` é injetado
 *  no api.ts (ignorado pelo backend, que usa a sessão). */
export interface LancamentoInput {
  descricao: string;
  valor_total: string;
  conta_id: string;
  categoria_id?: string | null;
  data_competencia?: string | null;
  status?: 'paga' | 'prevista';
}

// ── Orçamentos ──────────────────────────────────────────────────────
export interface Orcamento {
  id: string;
  usuario_id: string;
  categoria_id: string;
  categoria_nome?: string | null;
  valor_mensal: string;
  ativo: boolean;
}

export interface OrcamentoListResponse {
  items: Orcamento[];
  total: number;
}

export interface OrcamentoStatusItem {
  orcamento_id: string;
  categoria_id: string;
  categoria_nome?: string | null;
  valor_mensal: string;
  consumido: string;
  restante: string;
  percentual: number;
}

export interface OrcamentoStatusResponse {
  competencia: string;
  items: OrcamentoStatusItem[];
  total_orcado: string;
  total_consumido: string;
}

export interface OrcamentoCreateInput {
  categoria_id: string;
  valor_mensal: string;
}

export interface OrcamentoUpdateInput {
  valor_mensal?: string;
  ativo?: boolean;
}

// ── Pagar o mês (boletos + faturas juntos) ─────────────────────────
export interface PagamentoMesItem {
  tipo: 'boleto' | 'fatura';
  id: string;
  descricao: string;
  valor: string;
  vencimento?: string | null;
  conta_sugerida_id?: string | null;
  conta_sugerida_nome?: string | null;
}

export interface PagamentoMesPreview {
  competencia: string;
  itens: PagamentoMesItem[];
  total: string;
}

export interface PagamentoMesItemInput {
  tipo: 'boleto' | 'fatura';
  id: string;
  conta_id: string;
}

export interface PagamentoMesResultado {
  pagos: number;
  total_pago: string;
  falhas: string[];
}

export interface NluInterpretacao {
  tipo: string;
  valor: string;
  descricao: string;
  data: string;
  conta_id?: string | null;
  conta_nome?: string | null;
  categoria_id?: string | null;
  categoria_nome?: string | null;
  texto_original: string;
}

export interface PagamentoInput {
  conta_id: string;
  valor: string;
}

export interface DespesaDivididaInput {
  descricao: string;
  valor_total: string;
  pagamentos: PagamentoInput[];
  categoria_id?: string | null;
  data_competencia?: string | null;
  status?: 'paga' | 'prevista';
}

export interface PixParse {
  valor?: string | null;
  beneficiario?: string | null;
  cidade?: string | null;
  chave?: string | null;
}

export interface DespesaAutoSplitInput {
  descricao: string;
  valor_total: string;
  conta_vr_id: string;
  conta_fallback_id: string;
  categoria_id?: string | null;
  data_competencia?: string | null;
  notas?: string | null;
}

/** Payload pra editar uma transação (PATCH). Inclui o `tipo` porque a edição
 *  pode trocar despesa↔receita. */
export interface TransacaoEditInput {
  tipo: 'despesa' | 'receita';
  descricao: string;
  valor_total: string;
  conta_id: string;
  categoria_id?: string | null;
  data_competencia?: string | null;
  status?: 'paga' | 'prevista';
}

// ── Recorrências (despesas/receitas fixas) ──────────────────────────
export interface Recorrencia {
  id: string;
  usuario_id: string;
  descricao: string;
  tipo: string;
  valor_estimado: string;
  dia_vencimento: number;
  frequencia: string;
  categoria_id?: string | null;
  conta_id?: string | null;
  forma_pagamento: string;
  cartao_id?: string | null;
  ativa: boolean;
}

export interface RecorrenciaStatusItem {
  recorrencia_id: string;
  descricao: string;
  forma_pagamento: string;
  valor_estimado: string;
  dia_vencimento: number;
  cartao_id?: string | null;
  situacao: 'paga' | 'prevista' | 'atrasada' | 'lancada_cartao' | 'nenhuma';
  transacao_id?: string | null;
  compra_id?: string | null;
}

export interface RecorrenciaStatusResponse {
  competencia: string;
  items: RecorrenciaStatusItem[];
}

export interface PagarMesInput {
  competencia?: string | null;
  conta_id?: string | null;
  data_pagamento?: string | null;
  valor_pago?: string | null;
}

export interface RecorrenciaListResponse {
  items: Recorrencia[];
  total: number;
}

export type FormaPagamento = 'conta' | 'cartao' | 'boleto';

export interface RecorrenciaCreateInput {
  descricao: string;
  tipo: 'despesa' | 'receita';
  valor_estimado: string;
  dia_vencimento: number;
  categoria_id?: string | null;
  conta_id?: string | null;
  forma_pagamento?: FormaPagamento;
  cartao_id?: string | null;
}

export interface RecorrenciaUpdateInput {
  descricao?: string;
  tipo?: 'despesa' | 'receita';
  valor_estimado?: string;
  dia_vencimento?: number;
  categoria_id?: string | null;
  conta_id?: string | null;
  forma_pagamento?: FormaPagamento;
  cartao_id?: string | null;
  ativa?: boolean;
}

// ── Categorias ──────────────────────────────────────────────────────
export interface CategoriaTreeItem {
  id: string;
  nome: string;
  ativa: boolean;
  filhos: CategoriaTreeItem[];
}

export interface CategoriaTreeResponse {
  items: CategoriaTreeItem[];
  total: number;
}

export interface CategoriaResponse {
  id: string;
  nome: string;
  categoria_pai_id?: string | null;
  ativa: boolean;
}

export interface CategoriaCreateInput {
  nome: string;
  categoria_pai_id?: string | null;
}

export interface CategoriaUpdateInput {
  nome?: string;
  categoria_pai_id?: string | null;
  ativa?: boolean;
}

// ── Resumo / Relatório / Projeção ───────────────────────────────────
export interface ResumoMes {
  ano: number;
  mes: number;
  total_receitas: string;
  total_despesas: string;
  saldo: string;
  por_categoria: CategoriaResumoItem[];
}

export interface RelatorioMesItem {
  ano: number;
  mes: number;
  total_receitas: string;
  total_despesas: string;
  saldo: string;
}

export interface RelatorioResponse {
  meses: RelatorioMesItem[];
  por_categoria: CategoriaResumoItem[];
  total_receitas: string;
  total_despesas: string;
  saldo: string;
  media_despesas: string;
}

export interface ProjecaoMes {
  ano: number;
  mes: number;
  saldo_atual: string;
  a_pagar: string;
  a_receber: string;
  estimativa_sobra: string;
}

// ── Cartões / Faturas / Compras ─────────────────────────────────────
export interface Cartao {
  id: string;
  usuario_id: string;
  nome: string;
  bandeira?: string | null;
  dia_fechamento: number;
  dia_vencimento: number;
  limite?: string | null;
  ativo: boolean;
}

export interface CartaoListResponse {
  items: Cartao[];
  total: number;
}

export interface CartaoCreateInput {
  nome: string;
  bandeira?: string | null;
  dia_fechamento: number;
  dia_vencimento: number;
  limite?: string | null;
}

export interface CartaoUpdateInput {
  nome?: string;
  bandeira?: string | null;
  dia_fechamento?: number;
  dia_vencimento?: number;
  limite?: string | null;
  ativo?: boolean;
}

export interface Fatura {
  id: string;
  cartao_id: string;
  mes_referencia: string;
  valor_total: string;
  vencimento: string;
  status: string;
}

export interface FaturasCartao {
  cartao_id: string;
  faturas: Fatura[];
  total_em_aberto: string;
  total_juros: string;
}

export interface FaturaExtratoItem {
  parcela_id: string;
  compra_id: string;
  descricao: string;
  numero: number;
  total_parcelas: number;
  valor: string;
  valor_juros: string;
  vencimento: string;
  categoria_id?: string | null;
  categoria_nome?: string | null;
}

export interface FaturaExtrato {
  fatura: Fatura;
  cartao_nome: string;
  itens: FaturaExtratoItem[];
  total_juros: string;
}

export interface Parcela {
  id: string;
  numero: number;
  total_parcelas: number;
  valor: string;
  tem_juros: boolean;
  valor_juros: string;
  vencimento: string;
  fatura_id?: string | null;
}

export interface Compra {
  id: string;
  usuario_id: string;
  cartao_id?: string | null;
  descricao: string;
  valor_total: string;
  total_parcelas: number;
  data_compra: string;
  origem: string;
  categoria_id?: string | null;
  parcelas: Parcela[];
}

export interface PagarFaturaInput {
  conta_id: string;
  data_pagamento?: string | null;
  valor_pago?: string | null;
  categoria_id?: string | null;
}

export interface ProjecaoMesItem {
  mes_referencia: string; // "YYYY-MM-01"
  total: string;
}

export interface ProjecaoFaturas {
  meses: ProjecaoMesItem[];
  total: string;
}

export interface CompraCategoriaSugestao {
  categoria_id?: string | null;
  categoria_nome?: string | null;
}

export interface CompraCreateInput {
  cartao_id: string;
  descricao: string;
  valor_total: string;
  total_parcelas: number;
  data_compra?: string | null;
  categoria_id?: string | null;
  valor_juros_total?: string;
}

// ── Leituras de consumo (água/gás/luz) ──────────────────────────────
export interface LeituraConsumo {
  id: string;
  usuario_id: string;
  tipo: string;
  mes_referencia: string;
  leitura_atual: string;
  leitura_anterior?: string | null;
  consumo?: string | null;
  valor?: string | null;
  transacao_id?: string | null;
}

export interface LeituraConsumoListResponse {
  items: LeituraConsumo[];
  total: number;
}

export interface LeituraCreateInput {
  tipo: 'agua' | 'gas' | 'luz';
  mes_referencia: string;
  leitura_atual: string;
  leitura_anterior?: string | null;
  consumo?: string | null;
  valor?: string | null;
}

// ── Comprovantes ────────────────────────────────────────────────────
export interface Comprovante {
  id: string;
  usuario_id: string;
  transacao_id?: string | null;
  tipo: string;
  bucket: string;
  arquivo_path: string;
  nome_original?: string | null;
  content_type?: string | null;
  tamanho?: number | null;
  hash: string;
  url?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface ComprovanteListResponse {
  items: Comprovante[];
  total: number;
}

// ── Importador de boleto (LLM multimodal) ─────────────────────────
export interface VerbaBoleto {
  descricao: string;
  valor: string | number;
}

export interface LeituraBoleto {
  tipo: string;
  leitura_atual?: string | number | null;
  leitura_anterior?: string | number | null;
  consumo?: string | number | null;
  valor?: string | number | null;
}

export interface BoletoExtraido {
  beneficiario?: string | null;
  vencimento?: string | null;
  valor_total: string | number;
  linha_digitavel?: string | null;
  verbas: VerbaBoleto[];
  leituras: LeituraBoleto[];
}

export interface ImportarBoletoResponse {
  success: boolean;
  conferido: boolean;
  duplicado?: boolean;
  mensagem: string;
  comprovante_id?: string | null;
  transacao_id?: string | null;
  extraido?: BoletoExtraido | null;
}
